src/lib/store-and-retry.js

```
@@ -0,0 +1,170 @@
+let _postResponse = null;
+// Number of retries before abandon.
+let defaultMaxRetries = 10;
+let maxRetries = null;
+// Time between attempts.
+const threeMinutesInMilliseconds = 180000;
+let defaultIntervalBetweenAttempts = threeMinutesInMilliseconds;
+let intervalBetweenAttempts = null;
+let storage = null;
+let isStorageWritable = null;
+let isStorageReadable = null;
+// Name for the key in the web storage.
+const storageUniqueIdentifier = "n-feedback-response";
+let attempts = null;
+let scheduledRetry = null;
+// Customise error for simple identification.
+function customError(errorDescription, error) {
+  return new Error(`n-feedback: ${errorDescription}`, error);
+}
+// Save response locally if the storage is supported and a response has not been stored yet.
+function saveResponse(surveyId, surveyData, surveyResponse, additionalData) {
+  if (
+    isStorageWritable &&
+    isStorageReadable &&
+    !storage.getItem(storageUniqueIdentifier)
+  ) {
+    try {
+      storage.setItem(
+        storageUniqueIdentifier,
+        JSON.stringify({ surveyId, surveyData, surveyResponse, additionalData })
+      );
+    } catch (error) {
+      // It shouldn't happen, but in case of error we log it.
+      throw customError(`unable to write on storage`, error);
+    }
+  }
+}
+
+async function tryToSendResponse(
+  surveyId,
+  surveyData,
+  surveyResponse,
+  additionalData
+) {
+  attempts++;
+  if (attempts > maxRetries) {
+    stopRetry();
+    return Promise.resolve(
+      `Reached the maximum amount of retries (${maxRetries})`
+    );
+  } else {
+    try {
+      await _postResponse(surveyId, surveyData, surveyResponse, additionalData);
+      stopAndClearLocalData();
+      return Promise.resolve(`Suceeded at attempts number ${attempts}`);
+    } catch (err) {
+      // Store data, if not already existing, in case we are not able to resolve the retries succesfully.
+      saveResponse(surveyId, surveyData, surveyResponse, additionalData);
```

In saveResponse we could throw an error and at the moment is not catched.